### PR TITLE
Fix build on master

### PR DESCRIPTION
**Pull request by DavideD**
This PR will fix the issues we have on master and make sure that the build works.

The only test I couldn't fix was `CascadeTest` and I've decided to ignore it for now.
@gbadner: Maybe you can have a look at it when this PR is merged?

My understanding is that some operations wheren't executed in the right order.
This PR requires this patch for ORM to work: https://github.com/hibernate/hibernate-orm/pull/3258

### CODE REVIEW

**Comment by gavinking** [comment]
@DavideD I don't understand what's going on here.

This PR seems to involve a commits unrelated to the build?

`CascadeTest` passes for me on `master`. What issue were you seeing with it?

**Comment by DavideD** [comment]
Interesting, it doesn't work for me but I will double check

**Comment by DavideD** [comment]
I don't know why it works for you but master doesn't work for me and @gbadner (Gail, can you confirm this?). We have a lot of NullPointer related to many to one associations in several different tests. That's why I sent this PR.

**Comment by gavinking** [comment]
I have no clue what could be happening with you guys.

Screen Shot 2020-02-25 at 16 20 20

**Comment by gavinking** [comment]
And in IntelliJ:

Screen Shot 2020-02-25 at 16 24 10

**Comment by DavideD** [comment]
Out of curiosity: Does this PR work for you?
I'm interested to see why we have this behaviour

**Comment by gavinking** [comment]
> Out of curiosity: Does this PR work for you?

I'm trying by I need to build `hibernate-core:5.4.13-SNAPSHOT` which is taking a while.

**Comment by gavinking** [comment]
Screen Shot 2020-02-25 at 16 46 15

One supposes that this compilation error is because you're using a patched version of core.

**Comment by DavideD** [comment]
yep

File: README.md [hunk 1]
```diff
@@ -58,12 +58,40 @@ If you also want to run the MySQL tests, ensure that MySQL is installed,
 and then type the following:
 
     mysql -uroot
-    create database hibernaterx;
+    create database hibernate-rx;
     create user "hibernate-rx" identified by 'hibernate-rx';
-    grant all on hibernaterx.* to "hibernate-rx";
+    grant all on hibernate-rx.* to "hibernate-rx";
```
Comment (gavinking) [inline README.md]: @DavideD This isn't right. MySQL can't have a database named `hibernate-rx`, at least not on Mac.

**Comment by gavinking** [comment]
@DavideD When I build everything and run the tests, I have two test failures that result from you having changed the database name to something that MySQL doesn't allow.

**Comment by DavideD** [comment]
MMMh... this is weird. OK, feel free to ignore the commit

File: README.md (hunk 1 above)
Comment (gavinking) [reply README.md]: Hrrrrm, there might be a way to do it with `mysqladmin` rather than `mysql` but I can't seem to quite get it right.

Comment (gavinking) [reply README.md]: OK, I was wrong. I figured it out.

File: README.md [hunk 2]
```diff
@@ -58,12 +58,40 @@ If you also want to run the MySQL tests, ensure that MySQL is installed,
...
+```
+  sudo podman run --rm -it --name HibernateTestingMariaDB -e MYSQL_ROOT_PASSWORD=hibernate-rx -e MYSQL_DATABASE=hibernate-rx -e MYSQL_USER=hibernate-rx -e MYSQL_PASSWORD=hibernate-rx -p 3306:3306 mysql
+```
+
+* MySQL Client
+
+```
+  sudo podman exec -it HibernateTestingMariaDB mysql -U hibernate-rx -p
+```
+
```
Comment (gavinking) [inline README.md]: @DavideD Please spare us the use of `podman`. I tried that way and it didn't work at all. It's *way* easier to install Postgres or MySQL using a package manager like `brew` or whatever.

**Comment by DavideD** [comment]
Cool. If it causes issues we can use `hibernaterx` everywhere instead.
I just want the defaults to be the same (unless there are good reason not
to).

**Comment by gavinking** [comment]
> Those are quite useful for people like me that don't want to figure out how
to install multiple datastores everytime they want to test something.

Well my issue with this is that:

1. If we're talking about MySQL and Postgres, they're *super* easy to install using a regular package manager. On my box it's `brew install mysql` and `brew install postgresql`. And that's it. I'm sure it's gotta be just as easy on Linux. (Windows I dunno.)
2. On the other hand, even after I `brew install podman`, the commands you've got there don't actually work for me, because apparently additional setup is needed on MacOS. I don't think we need to have stuff that is so Linux-specific in the readme.
3. If we're talking about DB2 / Oracle / SQL Server, is it really easy to run these on podman?

Now, I get that it's nice how you can create a test user/database in the command you're using to start the database but really I don't see that as a big deal compared to figuring out how to make podman work.

> Would it be OK to move them in a separate file and just add a link to this new file in the README?

Sure, no problem.

**Comment by gavinking** [comment]
> Those are quite useful for people like me that don't want to figure out how
to install multiple datastores everytime they want to test something.

Well my issue with this is that:

1. If we're talking about MySQL and Postgres, they're *super* easy to install using a regular package manager. On my box it's `brew install mysql` and `brew install postgresql`. And that's it. I'm sure it's gotta be just as easy on Linux. (Windows I dunno.)
2. On the other hand, even after I `brew install podman`, the commands you've got there don't actually work for me, because apparently additional setup is needed on MacOS. I don't think we need to have stuff that is so Linux-specific in the readme.
3. If we're talking about DB2 / Oracle / SQL Server, is it really easy to run these on podman?

Now, I get that it's nice how you can create a test user/database in the command you're using to start the database but really I don't see that as a big deal compared to figuring out how to make podman work.

> Would it be OK to move them in a separate file and just add a link to this new file in the README?

Sure, no problem.

File: README.md (hunk 1 above)
Comment (gavinking) [reply README.md]: OK, @DavideD, I figured out what I was doing wrong. In the end you just need to quote the database name like [this](https://github.com/hibernate/hibernate-rx/commit/9122e4dba8d7915e8512ba0770a5d576b7e0e233).

But note that the code you have above won't work, at least not on Mac, but I would be surprised if it worked on any platform.

**Comment by gavinking** [comment]
> So obviously one difference between our setups is that I was using an older build of hibernate-core.

Ok, fine, so sorry, yep, after changing `hibernate-rx` to use Hibernate ORM 5.4.13-SNAPSHOT instead of 5.4.11-SNAPSHOT, I see the test failures you guys are talking about.

So what was it that changed in Hibernate ORM that required such a massive fix?!

**Comment by gbadner** [comment]
@DavideD, @gavinking, I'm still seeing 16 failures:

> Task :hibernate-rx-core:test

org.hibernate.rx.SingleTableInheritanceTest > testRootClassViaAssociation FAILED
    java.util.concurrent.CompletionException
        Caused by: java.lang.NullPointerException at SingleTableInheritanceTest.java:60

org.hibernate.rx.BasicTypesAndCallbacksTest > testBasicTypes FAILED
    java.util.concurrent.CompletionException
        Caused by: java.lang.NullPointerException at BasicTypesAndCallbacksTest.java:58

org.hibernate.rx.CascadeTest > testCascade FAILED
    java.util.concurrent.CompletionException
        Caused by: java.lang.NullPointerException at CascadeTest.java:43

org.hibernate.rx.MySQLBasicTest > testBasicTypes FAILED
    org.hibernate.service.spi.ServiceException
        Caused by: org.hibernate.exception.GenericJDBCException
            Caused by: java.sql.SQLException
    java.lang.NullPointerException

org.hibernate.rx.AutoincrementTest > testBasicTypes FAILED
    java.util.concurrent.CompletionException
        Caused by: java.lang.NullPointerException at AutoincrementTest.java:58

org.hibernate.rx.UnionSubclassInheritanceTest > testRootClassViaAssociation FAILED
    java.util.concurrent.CompletionException
        Caused by: java.lang.NullPointerException at UnionSubclassInheritanceTest.java:36

org.hibernate.rx.EagerOneToManyAssociationTest > findBookWithAuthors FAILED
    java.util.concurrent.CompletionException
        Caused by: java.lang.NullPointerException at EagerOneToManyAssociationTest.java:64

org.hibernate.rx.JoinedSubclassInheritanceTest > testRootClassViaAssociation FAILED
    java.util.concurrent.CompletionException
        Caused by: java.lang.NullPointerException at JoinedSubclassInheritanceTest.java:36

org.hibernate.rx.SecondaryTableTest > testRootClassViaAssociation FAILED
    java.util.concurrent.CompletionException
        Caused by: java.lang.NullPointerException at SecondaryTableTest.java:35

org.hibernate.rx.EagerManyToOneAssociationTest > persistTwoAuthors FAILED
    java.util.concurrent.CompletionException
        Caused by: java.lang.NullPointerException at EagerManyToOneAssociationTest.java:63

org.hibernate.rx.EagerManyToOneAssociationTest > persistOneBook FAILED
    java.util.concurrent.CompletionException
        Caused by: java.lang.NullPointerException at EagerManyToOneAssociationTest.java:34

org.hibernate.rx.EagerOneToOneAssociationTest > testPersist FAILED
    java.util.concurrent.CompletionException
        Caused by: java.lang.NullPointerException at EagerOneToOneAssociationTest.java:35

org.hibernate.rx.MySQLAutoincrementTest > testBasicTypes FAILED
    org.hibernate.service.spi.ServiceException
        Caused by: org.hibernate.exception.GenericJDBCException
            Caused by: java.sql.SQLException
    java.lang.NullPointerException

org.hibernate.rx.LazyManyToOneAssociationTest > fetchWithOneAuthor FAILED
    java.util.concurrent.CompletionException
        Caused by: java.lang.NullPointerException at LazyManyToOneAssociationTest.java:64

org.hibernate.rx.LazyManyToOneAssociationTest > fetchWithTwoAuthors FAILED
    java.lang.IllegalStateException
        Caused by: io.netty.channel.ChannelException
            Caused by: java.io.IOException

org.hibernate.rx.LazyManyToOneAssociationTest > fetchProfileWithOneAuthor FAILED
    org.hibernate.service.spi.ServiceException
        Caused by: java.lang.IllegalStateException
            Caused by: io.netty.channel.ChannelException
                Caused by: java.io.IOException

**Comment by gbadner** [comment]
Just deleted my comment with test failures, since I see you are in sync now.

**Comment by gbadner** [comment]
I'm trying to figure out what it was that caused the tests to start failing.

I'm having a hard time finding the right hibernate-rx / hibernate-orm commit combination that builds without all (most?) of those failures.

@gavinking, do you remember the commit at which your hibernate-rx branch was at when the tests were passing? Do you have any idea of the commit hibernate-orm was at, or at least the date you built hibernater-core.jar for testing hibernate-rx?

**Comment by gavinking** [comment]
@gbadner my best estimate would be right after we merged https://github.com/hibernate/hibernate-orm/pull/3176

HTH?

File: README.md (hunk 2 above)
Comment (Sanne) [reply README.md]: Could we keep a note for `podman` users? It's quite handy for those on Linux.

+1 for it to not be the only way

**Comment by Sanne** [comment]
> @gbadner my best estimate would be right after we merged [hibernate/hibernate-orm#3176](https://github.com/hibernate/hibernate-orm/pull/3176)
> 
> HTH?

That one was merged a long time ago?

**Comment by gavinking** [comment]
> That one was merged a long time ago?

Sure, but the build for Hibernate RX doesn't rebuild core, so if a change in core breaks RX, one doesn't  immediately notice. In this particular case, the problem was compounded by the fact that core *updated its version number* and RX was depending on the snapshot of an older version. So even rebuilding core didn't uncover the breakage.

**Comment by DavideD** [comment]
I'm going to close this PR and we can continue to discuss on the next one as necessary

**Pull request closed without merging**